mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMemorySearchHighlighter.java

```diff
@@ -0,0 +1,190 @@
+
+        SearchQuery searchQuery = expression.getSearchQuery();
+        ImmutableSet<MailboxId> inMailboxes = expression.getInMailboxes();
+        return Mono.fromCallable(() -> luceneMessageSearchIndex.searchDocument(inMailboxes, searchQuery, (int) limit)).flatMapMany(Flux::fromIterable)
+            .map(document -> Throwing.supplier(() -> buildSearchSnippet(document, searchQuery)).get());
+    }
+
+
+    public List<SearchSnippet> highlightSearchDirectly(MultimailboxesSearchQuery expression, MailboxSession session, long limit) {
+        BooleanQuery.Builder queryBuilder = new BooleanQuery.Builder();
+        //queryBuilder.add(buildQueryFromMailboxes(expression.getInMailboxes()), BooleanClause.Occur.MUST);
+        List<SearchQuery.Criterion> crits = expression.getSearchQuery().getCriteria();
+
+        if (crits.getFirst() instanceof SearchQuery.TextCriterion textCriterion) {
+            String searchText = textCriterion.getOperator().getValue().toUpperCase(Locale.US);
+
+            BooleanQuery subjectAndBodyQuery = new BooleanQuery.Builder()
+                .add(createTermQuery(BODY_FIELD, searchText), BooleanClause.Occur.SHOULD)
+                .add(createTermQuery(BASE_SUBJECT_FIELD, searchText), BooleanClause.Occur.SHOULD)
+                .build();
+            queryBuilder.add(subjectAndBodyQuery, BooleanClause.Occur.MUST);
+        }
+
+        BooleanQuery query = queryBuilder.build();
+
+        TopDocs topDocs = Throwing.supplier(() -> searcher.search(query, (int) limit)).get();
+
+        Highlighter highlighter = highlighter(query);
+        List<SearchSnippet> result = Stream.of(topDocs.scoreDocs)
+            .map(Throwing.function(sDoc -> searcher.storedFields().document(sDoc.doc)))
+            .map(document -> {
+                String subjectHighLighter = Throwing.supplier(() -> highlighter.getBestFragment(analyzer, BASE_SUBJECT_FIELD, document.get(BASE_SUBJECT_FIELD))).get();
+                String bodyHighLighter = Throwing.supplier(() -> highlighter.getBestFragment(analyzer, BODY_FIELD, document.get(BODY_FIELD))).get();
+                String searchSnippetId = String.valueOf(document.getField(UID_FIELD).numericValue().longValue());
+                return new SearchSnippet(searchSnippetId, subjectHighLighter, bodyHighLighter);
+            }).toList();
+
+        return result;
+    }
+
+    private Query buildQueryFromMailboxes(Collection<MailboxId> mailboxIds) {
+        BooleanQuery.Builder queryBuilder = new BooleanQuery.Builder();
+        for (MailboxId id : mailboxIds) {
+            String idAsString = id.serialize();
+            queryBuilder.add(new TermQuery(new Term(MAILBOX_ID_FIELD, idAsString)), BooleanClause.Occur.SHOULD);
+        }
+        return queryBuilder.build();
+    }
+
+    private Query createTermQuery(String fieldName, String value) {
+        if (suffixMatch) {
+            return new WildcardQuery(new Term(fieldName, "*" + value + "*"));
```

Comment:
Maybe not related to your changes but it could be vulnerable to a ReDos (Regex denial of service) as an adversary might control both the indexed document along with the query.

CF https://github.com/opensearch-project/OpenSearch/issues/687

BTW is there a code path where suffix match is `true`? Is it even relevant to keep this **dangerous** and **inneficient** code even alive?

Reply:
Ok a quick test

```
    @Test
    void testRedos() throws Exception {
        Map<String, String> headersTest = new HashMap<>();
        headersTest.put("Test", "test");
        headersTest.put("From", "test1 <test1@localhost>");
        headersTest.put("To", "test3 <test3@localhost>, test4 <test4@localhost>");
        headersTest.put("Cc", "test21 <test21@localhost>, test6 <test6@foobar>");
        MessageUid uid3 = MessageUid.of(38);
        MessageId id3 = factory.generate();
        Calendar cal = Calendar.getInstance();
        cal.set(1980, 2, 10);
        MessageBuilder builder3 = new MessageBuilder()
            .headers(headersTest)
            .flags(new Flags(Flag.DELETED))
            .mailboxId(TEST_ID_1)
            .uid(uid3)
            .internalDate(cal.getTime())
            .body(("a".repeat(2048) + "s").getBytes(UTF_8))
            .size(20);
        index.add(session, mailbox, builder3.build(id3)).block();

        SearchQuery query = SearchQuery.of(SearchQuery.bodyContains(".*a".repeat(512) + "c"));
        assertThat(index.search(session, mailbox, query).toStream())
            .isEmpty();
    }
```

proves that the code no longer is vulnerable...

```
org.apache.lucene.util.automaton.TooComplexToDeterminizeException: Determinizing automaton with 2564 states and 3589 transitions would require more than 10000 effort.

	at org.apache.lucene.util.automaton.Operations.determinize(Operations.java:735)
	at org.apache.lucene.util.automaton.CompiledAutomaton.<init>(CompiledAutomaton.java:195)
	at org.apache.lucene.search.AutomatonQuery.<init>(AutomatonQuery.java:124)
	at org.apache.lucene.search.AutomatonQuery.<init>(AutomatonQuery.java:100)
	at org.apache.lucene.search.AutomatonQuery.<init>(AutomatonQuery.java:83)
	at org.apache.lucene.search.AutomatonQuery.<init>(AutomatonQuery.java:69)
	at org.apache.lucene.search.WildcardQuery.<init>(WildcardQuery.java:50)
	at org.apache.james.mailbox.lucene.search.LuceneMessageSearchIndex.createTermQuery(LuceneMessageSearchIndex.java:876)
	at org.apache.james.mailbox.lucene.search.LuceneMessageSearchIndex.createTextQuery(LuceneMessageSearchIndex.java:1134)
	at org.apache.james.mailbox.lucene.search.LuceneMessageSearchIndex.createQuery(LuceneMessageSearchIndex.java:1222)
	at org.apache.james.mailbox.lucene.search.LuceneMessageSearchIndex.searchMultimap(LuceneMessageSearchIndex.java:521)
	at org.apache.james.mailbox.lucene.search.LuceneMessageSearchIndex.doSearch(LuceneMessageSearchIndex.java:488)
	at org.apache.james.mailbox.store.search.ListeningMessageSearchIndex.lambda$search$2(ListeningMessageSearchIndex.java:91)
```

Reply:
The linked ticket says "it's fixed" and the test demonstrate this is true